zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,244 @@
+//! A type that can hold the Zcash four types of value pools
+
+use crate::amount::{Amount, Constraint, Error, NegativeAllowed, NonNegative};
+
+/// A structure that hold amounts for the four value pool types.
+#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(bound = "C: Constraint")]
+pub struct ValueBalance<C = NegativeAllowed> {
+    transparent: Amount<C>,
+    sprout: Amount<C>,
+    sapling: Amount<C>,
+    orchard: Amount<C>,
+}
+
+impl<C> ValueBalance<C>
+where
+    C: Constraint + Copy,
+{
+    /// [Consensus rule]: The remaining value in the transparent transaction value pool MUST
+    /// be nonnegative.
+    ///
+    /// This rule applies to Block and Mempool transactions.
+    ///
+    /// [Consensus rule]: https://zips.z.cash/protocol/protocol.pdf#transactions
+    pub fn remaining_transaction_value(&self) -> Result<Amount<NonNegative>, Error> {
+        // This rule checks the transparent value balance minus the sum of the sprout,
+        // sapling, and orchard value balances in a transaction is nonnegative.
+        (self.transparent - (self.sprout + self.sapling + self.orchard)?)?
+            .constrain::<NonNegative>()
+    }
+
+    /// Creates a `ValueBalance` from the given transparent amount.
+    pub fn from_transparent_amount(transparent_amount: Amount<C>) -> Self {
+        ValueBalance {
+            transparent: transparent_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given sprout amount.
+    pub fn from_sprout_amount(sprout_amount: Amount<C>) -> Self {
+        ValueBalance {
+            sprout: sprout_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given sapling amount.
+    pub fn from_sapling_amount(sapling_amount: Amount<C>) -> Self {
+        ValueBalance {
+            sapling: sapling_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Creates a `ValueBalance` from the given orchard amount.
+    pub fn from_orchard_amount(orchard_amount: Amount<C>) -> Self {
+        ValueBalance {
+            orchard: orchard_amount,
+            ..Default::default()
+        }
+    }
+
+    /// Insert a transparent value balance into a given `ValueBalance`
```

Comment:
Documentation: Let's say what happens to the other values here.

Reply:
Confusing names: Can you change the names of the `from_(pool)_value_balance` functions?

`from` is reserved for conversion functions:
https://rust-lang.github.io/api-guidelines/naming.html#casing-conforms-to-rfc-430-c-case

But these functions actually set a specific value, and leave the others alone.
Try to use a similar name to other Zebra code, or the Rust standard library.
(Search for `&mut` function arguments in our code!)

Reply:
Thanks for this change.

`transparent_value_balance` looks like a getter for `ValueBalance.transparent`:
https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter

Can we use names like `copy_transparent_value_balance` instead?

Reply:
Maybe @jvff has a better suggestion?

Reply:
I'd suggest adding getter and setter methods. It would provide ergonomics that are similar to what's here while also becoming a bit more predictable. Maybe something like:

```
pub fn transparent_amount(&self) -> Amount<C> {
    self.transporent
}

pub fn set_transparent_amount(&mut self, transparent: Amount<C>) -> &mut Self {
    self.transparent = transparent;
    self
}

// ...
```

Reply:
Thanks, i implemented @jvff suggestion in https://github.com/ZcashFoundation/zebra/pull/2505/commits/1f960c7cabc6508c1953d9ed04ff6d47b9d769a5